### ISSUE TITLE
Why the shared requirement on async*

### ISSUE THREAD

**Issue by nomeata** [dfinity/motoko#3892]
The `async*` blocks are merely about abstracting over stuff that calls, so I wonder why we still have the `shared` requirement for their types. There is never going to be serialization happening at an `await*`, right?

**Comment by crusso** [comment]
Yeah, I'd like to get rid of that requirement for both async and async*, actually. And do them together to avoid confusion.

**Comment by crusso** [comment]
I was assuming just removing the requirement on `async*` wouldn't be useful enough, but perhaps I'm wrong. If you can already write useful abstractions (and use them!) with just `async*` maybe we should just relax type `async* <typ>`.

**Comment by crusso** [comment]
@nomeata So I've been thinking about how to relax this for `async` (since it works, modulo an arity bug, for `async*`).

Unfortunately, it's a little harder because of the way we currently compile shared functions. Instead of representing them as lambdas that send data and return an async, we actually represent shared functions as simple pairs of (principal x methodname) and, for applications, generate the marshalling code at the call site, based on the type of the function *at the call-site*. In a polymorphic context, this may not be not be a fully known type (i.e. involve type parameters).

My gut feeling is that, in order to relax the shared type constraint on `async <typ>`, we actually need to compile shared functions as proper or special lambdas, that encapsulate their argument serialization and result deserialization (where the types are known) at the definition site, so the call-sites can just be compiled to applications with no extra type info required. 
Unfortunately, this would also mean representing actors as records of functions, not just principals, and adjusting deserialization of actor and function references to introduce the required actor/function proxies, driven by the expected Motoko type during deserialization (I guess). 

For reference equality, we probably want shared functions to be slightly special lambdas, with fixed principal and method name fields, so we can test their equality by  field comparison (and perhaps even with a special GC tag).

Another variation would be to represent shared function not by full lambdas, but by 4-tuples, with the new fields carrying argument/result serialization/deserialization code, but I'm not sure that buys us much.

**Comment by nomeata** [comment]
> For reference equality, we probably want shared functions to be slightly special lambdas, with fixed principal and method name fields

Hmm, good point. And not just for equality, but also for passing them arguments to a shared function, i.e. for serialization?
